### PR TITLE
Refactor CheckedModeBanner into something more reusable.

### PR DESCRIPTION
**Pull request by apwilson**
@abarth

### CODE REVIEW

File: packages/flutter/lib/src/widgets/checked_mode_banner.dart [hunk 1]
```diff
@@ -7,58 +7,99 @@ import 'dart:math' as math;
...
+class _BannerPainter extends CustomPainter {
+  const _BannerPainter({this.message, this.location});
+
+  final String message;
+  final BannerLocation location;
 
   static const Color kColor = const Color(0xA0B71C1C);
-  static const double kOffset = 40.0; // distance to bottom of banner, at a 45 degree angle inwards from the top right corner
+  static const double kOffset =
+      40.0; // distance to bottom of banner, at a 45 degree angle inwards
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/checked_mode_banner.dart]: I'm not sure this is a formatting improvement.  :)

Comment (apwilson) [inline packages/flutter/lib/src/widgets/checked_mode_banner.dart]: haha figured you'd notice - ok, gimme a sec

**Comment by abarth** [comment]
This is really for @Hixie .  However, I'd recommend that you clean up the formatting a bit.

**Comment by apwilson** [comment]
Okay reformatted to be more inline with what it was - dart formatter on save is what's screwed it up.  

@Hixie 

btw, what's your stance when you see inconsistencies in a files formatting?  Leave them be?

**Comment by Hixie** [comment]
always fix all formatting to be in line with our style guide.

File: packages/flutter/lib/src/widgets/checked_mode_banner.dart [hunk 2]
```diff
@@ -66,12 +111,34 @@ class CheckedModeBanner extends StatelessWidget {
...
+        message: 'SLOW MODE',
+        location: BannerLocation.topRight);
       return true;
     });
     return result;
   }
 }
+
+/// Displays a banner.
+class Banner extends StatelessWidget {
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/checked_mode_banner.dart]: move this to above CheckedModeBanner, and rename the file to just banner.dart

File: packages/flutter/lib/src/widgets/checked_mode_banner.dart [hunk 3]
```diff
@@ -66,12 +111,34 @@ class CheckedModeBanner extends StatelessWidget {
   Widget build(BuildContext context) {
     Widget result = child;
     assert(() {
-      result = new CustomPaint(
-        foregroundPainter: const _CheckedModeBannerPainter(),
-        child: result
-      );
+      result = new Banner(
+        child: child,
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/checked_mode_banner.dart]: child: result

File: packages/flutter/lib/src/widgets/checked_mode_banner.dart [hunk 4]
```diff
@@ -38,21 +41,63 @@ class _CheckedModeBannerPainter extends CustomPainter {
...
+      case BannerLocation.topLeft:
+        return 0.0;
+    }
+  }
+
+  double get _rotation {
+    switch (location) {
+      case BannerLocation.bottomLeft:
+      case BannerLocation.topRight:
+        return math.PI / 4;
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/checked_mode_banner.dart]: 4.0

**Comment by Hixie** [comment]
now that this has some complex logic about positioning and so on, it would be nice for Banner to have a test.

File: packages/flutter/lib/src/widgets/checked_mode_banner.dart (hunk 3 above)
Comment (apwilson) [inline packages/flutter/lib/src/widgets/checked_mode_banner.dart]: I'm curious about this one; what's the difference?

File: packages/flutter/lib/src/widgets/checked_mode_banner.dart (hunk 2 above)
Comment (apwilson) [inline packages/flutter/lib/src/widgets/checked_mode_banner.dart]: done

File: packages/flutter/lib/src/widgets/checked_mode_banner.dart (hunk 4 above)
Comment (apwilson) [inline packages/flutter/lib/src/widgets/checked_mode_banner.dart]: done

File: packages/flutter/lib/src/widgets/checked_mode_banner.dart (hunk 3 above)
Comment (Hixie) [inline packages/flutter/lib/src/widgets/checked_mode_banner.dart]: If we later add another thing around this (like, say, ExcludeSemantics, which come to think maybe we should) then it won't be randomly buggy in one of the code paths.

Comment (apwilson) [inline packages/flutter/lib/src/widgets/checked_mode_banner.dart]: done

**Comment by apwilson** [comment]
done

File: packages/flutter/lib/src/widgets/banner.dart [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+      case BannerLocation.topRight:
+        return math.PI / 4.0;
+      case BannerLocation.bottomRight:
+      case BannerLocation.topLeft:
+        return -math.PI / 4.0;
+    }
+  }
+}
+
+/// Displays a banner.
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/banner.dart]: Remove the dartdoc unless you can make it much more detailed. :-)

**Comment by Hixie** [comment]
lgtm

**Pull request merged**